### PR TITLE
Stage 1

### PR DESCRIPTION
**Pull request by mihhu**
Implemented a web SDK that exposes the following functionality:

- A generic React component encapsulating the iframe API - `JitsiMeeting`
- Another React component that wraps the former, but it's auto configured for JaaS (the domain defaults to 8x8.vc, the tenant is required) - `JaaSMeeting`
- A framework agnostic way to dynamically load the external API (support fo JaaS included) - `initExternal`
- TypeScript types for JitsiMeetExternalAPI - `./types`

### CODE REVIEW

File: src/components/JitsiMeeting.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      }
+      setJitsiClass(JitsiClass);
+    });
+  }, []);
+
+  useEffect(() => {
+    if (jitsiClass) {
+      loadIFrame(jitsiClass);
+    }
+  }, [jitsiClass]);
```
Comment (quitrk) [inline src/components/JitsiMeeting.tsx:55]: `loadIFrame` should also be specified as a dependency

File: src/components/JitsiMeeting.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+
+  const loadIFrame = (JitsiMeetExternalAPI: any) => {
```
Comment (quitrk) [inline src/components/JitsiMeeting.tsx:58]: we should wrap this in a `useCallback`

File: src/components/JitsiMeeting.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    apiRef.current = new JitsiMeetExternalAPI.fn(domain, {
+      ...options,
+      roomName: getRoomName(tenant, options.roomName),
+      parentNode: meetingRef.current
+    });
+    setLoading(false);
+    onApiReady(apiRef.current);
+  };
+
+  const renderLoadingSpinner = () => {
```
Comment (quitrk) [inline src/components/JitsiMeeting.tsx:68]: ditto

File: src/components/JitsiMeeting.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        onApiReady={(externalApi) => console.log(externalApi)}
+      />
+  ```
+ */
+const JitsiMeeting = ({
+  id,
+  domain,
+  options,
+  tenant,
+  spinner,
```
Comment (quitrk) [inline src/components/JitsiMeeting.tsx:31]: it would seem that the `spinner` prop is expected to be directly a react component rather than a wrapper function over a react component. Maybe it would be better to pass it that way? wdyt?

File: src/init.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
+import { JaaS_DOMAIN } from "./constants";
+
+type ExternalApiMap = {
```
Comment (quitrk) [inline src/init.ts:3]: could you please explain the reasoning behind this? Do we expect multiple domains to be accessed by users during the lifecycle of an app?

File: src/init.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+          // cb can be `setJitsiClass` when called from component
+          // or the `resolve` function when from initExternal
+          cb(externalApiMap[domain].err);
+        });
+      }
+    };
+
+    document.head.appendChild(script as Node);
+  } else {
+    if (externalApiMap[domain].isLoaded) {
```
Comment (quitrk) [inline src/init.ts:75]: I can only see this being false following consecutive init calls in the same stack

File: src/init.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      externalApiMap[domain].isLoaded = true;
+      externalApiMap[domain].err = new Error(
+        `Script load error: ${script.src}`
+      );
+      const callbacks = externalApiMap[domain].callbacks;
+      if (callbacks && callbacks.length) {
+        callbacks.forEach((cb) => {
+          // cb can be `setJitsiClass` when called from component
+          // or the `resolve` function when from initExternal
+          cb(externalApiMap[domain].err);
```
Comment (quitrk) [inline src/init.ts:68]: any particular reason you opted here for a node style error first callback and for a promise based approach for the `initExternal`?

File: src/types/IJitsiMeetingOptions.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
+export default interface IJitsiMeetingOptions {
```
Comment (quitrk) [inline src/types/IJitsiMeetingOptions.ts:1]: it would seem that most of these are not being used. Are they simply for documentation purposes?

File: src/constants/index.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+export const DEFAULT_DOMAIN = "meet.jit.si";
+
+export const DEFAULT_OPTIONS = {
+  roomName: "DefaultRoomNameExample",
+  width: "100%",
+  height: "700px",
+};
+
+export const JaaS_DOMAIN = "8x8.vc";
```
Comment (quitrk) [inline src/constants/index.ts:9]: nit: inconsistent naming, I get why, just seems a bit weird to not have it all uppercase.

File: src/components/JaaSMeeting.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  ```
+ */
+const JaaSMeeting = ({
+  id,
+  options,
+  appId,
+  spinner,
+  onApiReady
+}: IJaaSMeetingProps) => {
+  const props = {
```
Comment (quitrk) [inline src/components/JaaSMeeting.tsx:32]: looks like you could just pass through the props and only add domain and tenant

File: .eslintrc.js [hunk 11]
```diff
@@ -0,0 +1,228 @@
+module.exports = {
```
Comment (quitrk) [inline .eslintrc.js:1]: any reason we're not using the extended configs as they are?

**Review by quitrk — CHANGES_REQUESTED** [review]

File: .eslintrc.js [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+  },
+  "ignorePatterns": ["**/__tests__/*"],
+  "plugins": [
+    "eslint-plugin-jsdoc",
+    "eslint-plugin-prefer-arrow",
+    "eslint-plugin-react",
+    "@typescript-eslint",
+    "jest"
+  ],
+  "rules": {
```
Comment (saghul) [inline .eslintrc.js:26]: Where do all these rules come from?

File: README.md [hunk 13]
```diff
@@ -1 +1,38 @@
...
+```
+### Modules
+#### initExternal
+To import the Jitsi Meet External API in a non-React project:
+```js
+// index.js
+initExternal().then((externalApi) => { ... });
+
+// index.html
+window.onload = () => {
```
Comment (saghul) [inline README.md:16]: Why would one do this? It can all be handled in the JS file, right? Once `initExternal` resolves one can start writing the app logic already.

File: babel.config.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
+module.exports = {
+  "env": {
+    "test": {
+      presets: [
+        ['@babel/preset-env', {targets: {node: 'current'}}],
```
Comment (saghul) [inline babel.config.js:5]: Node is not a target we care about, we target browsers only with this library.

File: examples/initExternal/index.html [hunk 15]
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
```
Comment (saghul) [inline examples/initExternal/index.html:1]: I'll skip the examples for this first review round and focus on the SDK code since they will need adjusting.

File: package.json [hunk 16]
```diff
@@ -0,0 +1,62 @@
+{
+  "name": "jitsi-meet-web-sdk",
```
Comment (saghul) [inline package.json:2]: Make this `@jitsi/web-sdk` since we'll likely release it as such.

File: README.md [hunk 17]
```diff
@@ -1 +1,38 @@
...
+  const api = new window.JitsiMeetExternalAPI("YOUR_DOMAIN", { roomName: "YOUR_CUSTOM_ROOM_NAME" });
```
Comment (horymury) [inline README.md:17]: If the initExternal() function is used, then I think here the instantiation should be done on the result of that promise and not with windows.JitsiMeetExternalAPI

File: package.json [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+  "version": "0.0.0",
+  "description": "Web SDK for the Jitsi Meet IFrame",
+  "main": "lib/index.js",
+  "types": "lib/index.d.ts",
+  "scripts": {
+    "build": "tsc",
+    "format": "prettier --write \"src/**/*.ts\" \"src/**/*.js\"",
+    "lint": "eslint -c .eslintrc.js --ext .ts \"src\"",
+    "postversion": "git push && git push --tags",
```
Comment (saghul) [inline package.json:11]: Please drop this, raising the version should not push automatically. We'll use PRs.

File: src/components/JitsiMeeting.tsx [hunk 19]
```diff
@@ -0,0 +1,90 @@
+import { useEffect, useRef, useState } from "react";
```
Comment (saghul) [inline src/components/JitsiMeeting.tsx:1]: Please use our eslint base: https://github.com/jitsi/eslint-config-jitsi which has the main rules we follow. Single quotes being one of them ;-)

File: src/components/JitsiMeeting.tsx [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Returns the JitsiMeeting Component with access to a custom External API
+ * to be used as-it-is in React projects
+ *
+ * @param {IJitsiMeetingProps} props
+ * @example
+  ```js
+      <JitsiMeeting
+        id="testId"
```
Comment (saghul) [inline src/components/JitsiMeeting.tsx:14]: What is this ID?

File: src/init.ts [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+  new Promise((resolve, reject) => {
+    init(domain, (error: Error | null, jitsiClass?: any): void => {
+      if (error) {
+        reject(error);
+      }
+      resolve(jitsiClass);
+    });
+  });
+
+export const init = (
```
Comment (saghul) [inline src/init.ts:33]: Please give this a more descriptive name such as `initExternalAPI` and make the default domain be `meet.jit.si`.

File: src/init.ts [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+  domain: string,
+  callback: (err: Error | null, jitsiClass?: any) => void
```
Comment (saghul) [inline src/init.ts:35]: Why not return a Promise?

File: src/init.ts [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+    externalApiMap[domain] = {
+      isLoaded: false,
+      callbacks: [callback],
+      err: null,
+    };
+    const script: HTMLScriptElement = document.createElement("script");
+    script.async = true;
+    script.src = `https://${domain}/external_api.js`;
+    script.onload = () => {
+      externalApiMap[domain].isLoaded = true;
```
Comment (saghul) [inline src/init.ts:47]: It doesn't really matter where the external_api.js file is loaded from, it can then be used to join meetings in any other domain, so we only need to have it loaded once.

File: src/init.ts [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+      externalApiMap[domain].fn = window.JitsiMeetExternalAPI.bind({});
```
Comment (saghul) [inline src/init.ts:48]: Why are we wrapping it like this, instead of returning `window.JitsiMeetExternalAPI` directly?

File: src/types/IJaaSMeetingProps.ts [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+export default interface IJaaSMeetingProps {
+  /**
+   * The id of the component
+   */
+  id: string;
+
+  /**
+   * The App ID that provides an isolated context and prefixes the room name
+   */
+  appId: string;
```
Comment (saghul) [inline src/types/IJaaSMeetingProps.ts:17]: Does this map to the tenant? How do we refer to this in the JaaS docs?

File: src/types/IJaaSMeetingProps.ts [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+import { JaaSMeeting } from "..";
+
+/**
+ * The type of the React {@code Component} props of {@link JaaSMeeting}.
+ */
+export default interface IJaaSMeetingProps {
+  /**
+   * The id of the component
+   */
+  id: string;
```
Comment (saghul) [inline src/types/IJaaSMeetingProps.ts:12]: Why is this needed? I wouldn't know what to set it to.

File: src/types/IJaaSMeetingProps.ts [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+
+  /**
+   * The App ID that provides an isolated context and prefixes the room name
+   */
+  appId: string;
+
+  /**
+   * The optional arguments for the conference
+   */
+  options: IJitsiMeetingOptions;
```
Comment (saghul) [inline src/types/IJaaSMeetingProps.ts:22]: I'd say we flatten the structure here and have all options be top level options.

File: src/types/IJaaSMeetingProps.ts [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+
+  /**
+   * The custom spinner to be displayed while the iframe is loading
+   */
+  spinner?: () => JSX.Element;
```
Comment (saghul) [inline src/types/IJaaSMeetingProps.ts:27]: Why a function and not a component, directly?

File: src/types/IJaaSMeetingProps.ts [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+
+  /**
+   * The external API reference for events and commands
+   */
+  onApiReady: (api: any) => object;
```
Comment (saghul) [inline src/types/IJaaSMeetingProps.ts:32]: Can you please add a types file for the external API, so this can be `JitsiMeetExternalAPIType` instead of `any` ?

File: src/types/IJitsiMeetingProps.ts [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+  id: string;
+  /**
+   * The domain used to build the conference URL
+   */
+  domain: string;
+
+  /**
+   * The tenant that provides an isolated context and prefixes the room name
+   */
+  tenant?: string;
```
Comment (saghul) [inline src/types/IJitsiMeetingProps.ts:21]: This doesn't apply to non jaas meetings.

File: tsconfig.json [hunk 31]
```diff
@@ -0,0 +1,20 @@
+{
+  "compilerOptions": {
+    "target": "es6",
+    "module": "commonjs",
```
Comment (saghul) [inline tsconfig.json:4]: We probably want this to be es6 or esnext.

File: src/types/IJitsiMeetingOptions.ts [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+  /**
+   * The name of the room to join
+   */
+  roomName?: string;
+
+  /**
+   * The created IFrame width
+   */
+  width?: string | number;
```
Comment (saghul) [inline src/types/IJitsiMeetingOptions.ts:10]: Since we are using the parentNode approach it probably doesn't make sense to expose these?

**Review by saghul — COMMENTED** [review]
Great start! Left some comments, PTAL. Can you please also add a GH action to run the tests?

File: src/types/IJitsiMeetingProps.ts (hunk 30 above)
Comment (horymury) [reply src/types/IJitsiMeetingProps.ts:21]: @saghul  we do have tenant notion outside of JaaS, e.g. WorkD users and 8x8 Meet users

Comment (saghul) [reply src/types/IJitsiMeetingProps.ts:21]: It's part of the room name really, not sure we want to expose the concept.

File: src/init.ts (hunk 24 above)
Comment (horymury) [reply src/init.ts:48]: Me and @mihhu  brainstormed this part, we did all this mapping and cloning of the JitsiMeetExternalAPI class since components can be on different domains. While generally the external_api.js can be used inter-domains, there are cases in which on one domain we have one or more commands/functions/events which are not found on external_api from other domain.

File: src/init.ts (hunk 23 above)
Comment (horymury) [reply src/init.ts:47]: > It doesn't really matter where the external_api.js file is loaded from, it can then be used to join meetings in any other domain, so we only need to have it loaded once.

Me and @mihhu  brainstormed this part, we did all this mapping and cloning of the JitsiMeetExternalAPI class since components can be on different domains. While generally the external_api.js can be used inter-domains, there are cases in which on one domain we have one or more commands/functions/events which are not found on external_api from other domain. But I do agree that this complicated things quite a bit.

File: src/init.ts (hunk 24 above)
Comment (saghul) [reply src/init.ts:48]: This has never happened IRL, so let's not over-complicate things unless we absolutely need to, which is not the case here :-)

File: src/init.ts (hunk 23 above)
Comment (saghul) [reply src/init.ts:47]: As I mentioned above, let's drop this for now.

File: src/types/IJitsiMeetingProps.ts (hunk 30 above)
Comment (horymury) [reply src/types/IJitsiMeetingProps.ts:21]: we do expose it already one way or another on our products. I personally don't see a reason why not to expose it.

Comment (saghul) [reply src/types/IJitsiMeetingProps.ts:21]: Not opposed, but it was never part of the external API itself and it does require setup on the infra side. The Docker setup, for example, does not support tenants. Not all Jitsi installations support tenants. That's why I thought of not exposing it.

File: src/types/IJaaSMeetingProps.ts (hunk 29 above)
Comment (horymury) [reply src/types/IJaaSMeetingProps.ts:32]: does this return `object`? or `void`?

File: src/types/IJitsiMeetingProps.ts [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+
+  /**
+   * The optional arguments for the conference
+   */
+  options: IJitsiMeetingOptions;
+
+  /**
+   * The external API reference for events and commands
+   */
+  onApiReady: (api: any) => object;
```
Comment (horymury) [inline src/types/IJitsiMeetingProps.ts:36]: ditto

File: package.json [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+    "test": "jest --config jestconfig.json",
+    "version": "npm run format && git add -A src"
```
Comment (horymury) [inline package.json:13]: maybe add a lint-fix script

**Review by horymury — CHANGES_REQUESTED** [review]

File: src/types/IJitsiMeetingOptions.ts (hunk 32 above)
Comment (mihhu) [reply src/types/IJitsiMeetingOptions.ts:10]: I think it's the parentNode that we shouldn't expose, because it's overwritten when the component is loading the IFrame, as the root of the component itself acts as a 'parent node'. What do you think?

File: src/types/IJitsiMeetingOptions.ts (hunk 8 above)
Comment (mihhu) [reply src/types/IJitsiMeetingOptions.ts:1]: Yep 😃

File: src/types/IJaaSMeetingProps.ts (hunk 25 above)
Comment (mihhu) [reply src/types/IJaaSMeetingProps.ts:17]: It does, but I'll remove the tenant prop as per the other threads.

File: src/types/IJaaSMeetingProps.ts (hunk 26 above)
Comment (mihhu) [reply src/types/IJaaSMeetingProps.ts:12]: Just to allow more customisation and to make sure the root of the instance is unique, to avoid other react related issues. Will keep it internal, maybe with the possibility of accessing it the same way we do the external API (if it's ok with you).

For the internal implementation, @horymury suggested the use of `generateGUID` from the 8x8meet core, would that be ok?

File: src/init.ts (hunk 6 above)
Comment (mihhu) [reply src/init.ts:75]: Exactly, but if you wish to add another instance later on, this part will come handy.

File: src/init.ts (hunk 5 above)
Comment (mihhu) [reply src/init.ts:3]: Yes, that was angle. But after all the feedback, it doesn't seem to ever be the case, so I'll change this part.

File: src/components/JitsiMeeting.tsx (hunk 2 above)
Comment (mihhu) [reply src/components/JitsiMeeting.tsx:58]: Given the nature of this component, I'm not entirely sure it will help, but rather make the first render take longer than needed, without getting any of the caching benefits later on. I might be wrong though, can you please explain why should I do this?

File: src/types/IJitsiMeetingOptions.ts (hunk 32 above)
Comment (saghul) [reply src/types/IJitsiMeetingOptions.ts:10]: I thought we'd always render our stuff inside a div, and use that to contain the spinner too. Is there a downside to this?

Comment (mihhu) [reply src/types/IJitsiMeetingOptions.ts:10]: I'm not sure I got that. The spinner is rendered inside the component and next to the only div that acts as a parent node for the IFrame. The width and height are still needed as props, because we don't want to 'hardcode' them in the component logic, is that ok?

Comment (saghul) [reply src/types/IJitsiMeetingOptions.ts:10]: We should probably allow passing styles for the div we create or a class name. WDYT would make more sense here?

Comment (horymury) [reply src/types/IJitsiMeetingOptions.ts:10]: or we could just expose the ref of the container div from our component, and the integrator can add any styles/classes to it. wdyt?

Comment (saghul) [reply src/types/IJitsiMeetingOptions.ts:10]: How would customization work? Can you show a quick example?

Comment (horymury) [reply src/types/IJitsiMeetingOptions.ts:10]: Was thinking something around the lines of:

Integrator:
```
const getMeetingDivRef  = meetingDivRef => {
   integratorMeetingRef = meetingDivRef; // integrator preserves the ref in their app.
  /// do whatver we want with the ref, including re-style
  meetingDivRef.style.backgroundColor = "#FFF';
  meetingDivRef.classList.add('integrator-class');
}

// Integrator instantiates the JitsiMeeting component with passing the callback function.
<JitsiMeeting ... getRef = {getMeetingDivRef} />
```


And inside the component once everything is loaded, we already have the `meetingRef` as the reference to the meeting's container div so we call (maybe inside the `loadIframe` function):
```
 getRef && getRef(meetingRef.current);
```

File: src/components/JitsiMeeting.tsx (hunk 2 above)
Comment (quitrk) [reply src/components/JitsiMeeting.tsx:58]: You are right, we probably over do it on other projects for consistency with actual callbacks and for the sake of being explicit about when a function should be redeclared. I'm not sure I agree about the rendering taking longer though.

File: src/types/IJaaSMeetingProps.ts (hunk 29 above)
Comment (mihhu) [reply src/types/IJaaSMeetingProps.ts]: @horymury `void`, changed accordingly, ty!

File: src/init.ts (hunk 22 above)
Comment (mihhu) [reply src/init.ts]: To be able to wait for the script to load before using the window object when multiple instances of the IFrame are rendered at the same time.

File: src/components/JitsiMeeting.tsx (hunk 20 above)
Comment (mihhu) [reply src/components/JitsiMeeting.tsx]: I needed some uniqueness when testing the multiple instance scenario. Had it removed and generated an uuid in the function component body.

File: .eslintrc.js (hunk 12 above)
Comment (mihhu) [reply .eslintrc.js:26]: `eslint --init` :D removed most of them, I have yet to figure out some issues with estlint-config-jitsi

File: src/init.ts (hunk 7 above)
Comment (mihhu) [reply src/init.ts]: To be able to wait for the script to load before using the window object when multiple instances of the IFrame are rendered at the same time (and without adding multiple identical script tags).

File: src/types/IJitsiMeetingOptions.ts (hunk 32 above)
Comment (mihhu) [reply src/types/IJitsiMeetingOptions.ts]: Implemented @horymury suggestion. @saghul, thoughts?

File: src/types/IJaaSMeetingProps.ts [hunk 35]
```diff
@@ -102,4 +102,9 @@ export default interface IJaaSMeetingProps {
    * The external API reference for events and commands
    */
   onApiReady: (api: IJitsiMeetExternalApi) => void;
+
+  /**
+   * The reference to the div used by the IFrame.
+   */
+   getIFrameRef?: (ref: MutableRefObject<HTMLElement>) => void;
```
Comment (horymury) [inline src/types/IJaaSMeetingProps.ts:109]: think the param is simple HTMLElement, as we pass `.current` to the method.

File: src/components/JitsiMeeting.tsx [hunk 36]
```diff
@@ -36,6 +36,7 @@ const JitsiMeeting = ({
   userInfo,
   spinner,
```
Comment (horymury) [inline src/components/JitsiMeeting.tsx:37]: you could use aliasing to uppercase spinner in the component's params instead of creating a new const on the spinner render function : {...,spinner: Spinner, ...} .Think it's cleaner this way.

File: src/types/IJaaSMeetingProps.ts [hunk 37]
```diff
@@ -0,0 +1,110 @@
+import React from 'react';
+
+/* eslint-disable-next-line */
+import { JaaSMeeting } from '..';
+import IJitsiMeetExternalApi from './IJitsiMeetExternalApi';
+
+/**
+ * The type of the React {@code Component} props of {@link JaaSMeeting}.
+ */
+export default interface IJaaSMeetingProps {
```
Comment (saghul) [inline src/types/IJaaSMeetingProps.ts:10]: THis looks 99% similar to the other one, maybe we can define a base interface from which both inherit?

File: src/init.ts [hunk 38]
```diff
@@ -0,0 +1,85 @@
...
+ * @returns {Promise<object>} - Object containing the JitsiMeetExternalAPI object
+ * or an error
+ */
+export const fetchExternalApi = (
+        domain: string = JAAS_DOMAIN
+): Promise<void | (new () => IJitsiMeetExternalApi)> =>
+    new Promise((resolve, reject) => {
+        initExternalApi(domain, (error: Error | null, api?: new () => IJitsiMeetExternalApi): void => {
+            if (error) {
+                reject(error);
```
Comment (saghul) [inline src/init.ts:81]: Unless you `return` here you are going to first call reject and then resolve right after.

File: src/init.ts [hunk 39]
```diff
@@ -0,0 +1,85 @@
...
+ * Injects the external_api.js script for the corresponding domain in DOM
+ * and resolves with an object containing either the `JitsiMeetExternalAPI`
+ * class definition or an error
+ *
+ * @param {string} domain - The domain of the external API
+ * @returns {Promise<object>} - Object containing the JitsiMeetExternalAPI object
+ * or an error
+ */
+export const fetchExternalApi = (
+        domain: string = JAAS_DOMAIN
```
Comment (saghul) [inline src/init.ts:76]: Why does this default to the JaaS domain?

File: src/init.ts [hunk 40]
```diff
@@ -0,0 +1,85 @@
...
+
+type ExternalApi = {
+    isLoaded: boolean;
+    callbacks: ((err: Error | null, api?: new () => IJitsiMeetExternalApi) => void)[];
+    err: Error | null;
+};
+
+let externalApi: ExternalApi;
+
+export const initExternalApi = (
```
Comment (saghul) [inline src/init.ts:12]: I still find this more complex than it needs to be. Maybe we can borrow some inspiration from https://github.com/feross/load-script2 ?

I think we can just have some state booleans to tell us if we are loading the external API, in which case we could chain a promise, or if it's loaded already, etc.

Can you give it another go?

File: src/constants/index.ts [hunk 41]
```diff
@@ -0,0 +1,9 @@
+export const DEFAULT_DOMAIN = 'meet.jit.si';
+
+export const DEFAULT_OPTIONS = {
+    roomName: 'DefaultRoomNameExample',
+    width: '100%',
```
Comment (saghul) [inline src/constants/index.ts:5]: Let's drop width and height, the iframe should cover 100% of the container we create and then the embedder can customize the container with the ref.

File: src/constants/index.ts [hunk 42]
```diff
@@ -0,0 +1,9 @@
+export const DEFAULT_DOMAIN = 'meet.jit.si';
+
+export const DEFAULT_OPTIONS = {
+    roomName: 'DefaultRoomNameExample',
```
Comment (saghul) [inline src/constants/index.ts:4]: Don't set a default room name please, people testing this may end up meeting strangers :-)

File: src/components/JitsiMeeting.tsx [hunk 43]
```diff
@@ -0,0 +1,126 @@
...
+                /* eslint-disable-next-line */
+                console.error(err);
+
+                return;
+            }
+            setExternalApi(api);
+        });
+    }, []);
+
+    const loadIFrame = useCallback((result: any) => {
```
Comment (saghul) [inline src/components/JitsiMeeting.tsx:60]: can we annotate the type of `result` please?

File: src/components/JitsiMeeting.tsx [hunk 44]
```diff
@@ -0,0 +1,126 @@
...
+ */
+const JitsiMeeting = ({
+    domain,
+    roomName,
+    width,
+    height,
+    configOverwrite,
+    interfaceConfigOverwrite,
+    jwt,
+    onload,
```
Comment (saghul) [inline src/components/JitsiMeeting.tsx:33]: I wonder if we want to expose the `onload` handler...

File: package.json [hunk 45]
```diff
@@ -0,0 +1,55 @@
...
+        "lint-fix": "npm run lint -- --fix",
+        "test": "jest --config jestconfig.json"
+    },
+    "repository": {
+        "type": "git",
+        "url": "git+https://github.com/jitsi/jitsi-meet-web-sdk.git"
+    },
+    "keywords": [],
+    "author": "",
+    "license": "Apache-2.0",
```
Comment (saghul) [inline package.json:19]: Please add a LICENSE file with Apache 2 too please.

**Review by saghul — COMMENTED** [review]
Left some more comments. PTAL! Let's land version 0.0.1 after those are addressed and move forward, we can make further improvements later.

File: src/init.ts (hunk 39 above)
Comment (mihhu) [reply src/init.ts:76]: Thought this would cover the last bullet of the requirements:

> A framework agnostic way to dynamically load the external API (support fo JaaS included)

Should I remove it?

Comment (saghul) [reply src/init.ts:76]: I'd say we either default to meet.jit.si or nothing at all here.

File: package.json (hunk 45 above)
Comment (mihhu) [reply package.json:19]: Isn't [this](https://github.com/jitsi/jitsi-meet-web-sdk/blob/main/LICENSE) enough?

File: src/init.ts [hunk 46]
```diff
@@ -1,17 +1,17 @@
 import { DEFAULT_DOMAIN, JAAS_DOMAIN } from './constants';
-import IJitsiMeetExternalApi from './types/IJitsiMeetExternalApi';
+import { JitsiMeetExternalApi } from './types';
 
 type ExternalApi = {
     isLoaded: boolean;
-    callbacks: ((err: Error | null, api?: new () => IJitsiMeetExternalApi) => void)[];
+    callbacks: ((err: Error | null, api?: JitsiMeetExternalApi) => void)[];
```
Comment (horymury) [inline src/init.ts:6]: I would name this JitsiMeetApi or even JitsiMeetExternalApi instead of `api`. We conventionally name `api` the instance, not the class definition.

File: package.json (hunk 45 above)
Comment (saghul) [reply package.json:22]: LoL I forgot I added that!

File: src/components/JitsiMeeting.tsx (hunk 44 above)
Comment (mihhu) [reply src/components/JitsiMeeting.tsx:29]: Not sure, what reasons are there not to? Maybe some integrators could use it to load extra functionality.

Comment (saghul) [reply src/components/JitsiMeeting.tsx:29]: It seems to overlap with the spinner. That exists so users know when to display some loading indication. But id we provide it I don't see its usefulness.

As a general note, it's always easier to extend an API than to take things back ;-)

Comment (mihhu) [reply src/components/JitsiMeeting.tsx:29]: Fair enough 😃

**Review by horymury — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 47]
```diff
@@ -19,4 +19,9 @@ jobs:
       - name: Build project
         run: npm run build
       - name: Run unit tests
-        run: npm test
\ No newline at end of file
+        run: npm test
+      - name: Release to npm
+        if: github.event_name == 'release' && github.event.action == 'published' && startsWith(matrix.os, 'macos')
```
Comment (saghul) [inline .github/workflows/ci.yml:24]: remove the last part, we are not testing on macos.

**Review by quitrk — APPROVED** [review]

**Comment by saghul** [comment]
Excellent work @mihhu !
